src/ImageSharp/Formats/Heif/HeifEncoderCore.cs

```diff
@@ -0,0 +1,334 @@
+    }
+
+    private static int WriteSpatialExtentPropertyBox(AutoExpandingMemory<byte> memory, int memoryOffset, HeifItem item)
+    {
+        Span<byte> buffer = memory.GetSpan(memoryOffset, 20);
+        int bytesWritten = WriteBoxHeader(buffer, Heif4CharCode.ispe, 0, 0);
+        BinaryPrimitives.WriteUInt32BigEndian(buffer[bytesWritten..], (uint)item.Extent.Width);
+        bytesWritten += 4;
+        BinaryPrimitives.WriteUInt32BigEndian(buffer[bytesWritten..], (uint)item.Extent.Height);
+        bytesWritten += 4;
+
+        BinaryPrimitives.WriteUInt32BigEndian(buffer, (uint)bytesWritten);
+        return bytesWritten;
+    }
+
+    private static int WriteItemDataBox(AutoExpandingMemory<byte> memory, int memoryOffset)
+    {
+        Span<byte> buffer = memory.GetSpan(memoryOffset, 10);
+        int bytesWritten = WriteBoxHeader(buffer, Heif4CharCode.idat);
+
+        BinaryPrimitives.WriteUInt32BigEndian(buffer, (uint)bytesWritten);
+        return bytesWritten;
+    }
+
+    private static int WriteItemLocationBox(AutoExpandingMemory<byte> memory, int memoryOffset, List<HeifItem> items)
+    {
+        Span<byte> buffer = memory.GetSpan(memoryOffset, 30 + (items.Count * 8));
+        int bytesWritten = WriteBoxHeader(buffer, Heif4CharCode.iloc, 1, 0);
+        buffer[bytesWritten++] = 0x44;
+        buffer[bytesWritten++] = 0;
+        BinaryPrimitives.WriteUInt16BigEndian(buffer[bytesWritten..], 1);
+        bytesWritten += 2;
+        BinaryPrimitives.WriteUInt16BigEndian(buffer[bytesWritten..], (ushort)items[0].Id);
+        bytesWritten += 2;
+        for (int i = 0; i < 4; i++)
+        {
+            buffer[bytesWritten++] = 0;
+        }
+
+        IEnumerable<HeifLocation> itemLocs = items.SelectMany(item => item.DataLocations).Where(loc => loc != null);
+        BinaryPrimitives.WriteUInt16BigEndian(buffer[bytesWritten..], (ushort)itemLocs.Count());
+        bytesWritten += 2;
+        foreach (HeifLocation loc in itemLocs)
+        {
+            BinaryPrimitives.WriteUInt32BigEndian(buffer[bytesWritten..], (uint)loc.Offset);
+            bytesWritten += 4;
+            BinaryPrimitives.WriteUInt32BigEndian(buffer[bytesWritten..], (uint)loc.Length);
+            bytesWritten += 4;
+        }
+
+        BinaryPrimitives.WriteUInt32BigEndian(buffer, (uint)bytesWritten);
+        return bytesWritten;
+    }
+
+    private void WriteMediaDataBox(Span<byte> data, Stream stream)
+    {
+        Span<byte> buf = stackalloc byte[12];
+        int bytesWritten = WriteBoxHeader(buf, Heif4CharCode.mdat);
+        BinaryPrimitives.WriteUInt32BigEndian(buf, (uint)(data.Length + bytesWritten));
+        stream.Write(buf[..bytesWritten]);
+
+        stream.Write(data);
+    }
+
+    private static async Task<byte[]> CompressPixels<TPixel>(Image<TPixel> image, CancellationToken cancellationToken)
+        where TPixel : unmanaged, IPixel<TPixel>
+    {
+        using MemoryStream stream = new();
+        JpegEncoder encoder = new()
```

Comment:
Is there any means to configure quality?

Reply:
Legacy JPEG is meant as a thumbnail format in HEIF. The primary formats is still to be added (Av1 in our case).
The current situation, with JPEG as main image is a bit strange, it would be more efficient to store the JPEG directly in its native format.

Av1 will have its own set of quality settings.

If you feel we need to set the quality of the thumbnail in the end situation, I'm happy to add that feature now already.

Reply:
No need then thanks.